File: sranco/sranco/report/snc_commision_history/snc_commision_history.js
// Copyright (c) 2023, Dinesh Panchal and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["SNC Commision History"] = {
  onload: function (report) {
    // Add fields for By Percent and By Amount
    report.page.add_field({
      fieldname: "by_percent_commission",
      label: __("By Commission Percent + / -"),
      fieldtype: "Float",
    });

    report.page.add_field({
      fieldname: "by_amount_commission",
      label: __("By Commission Amount + / -"),
      fieldtype: "Currency",
    });

    // Add Apply button
    report.page.add_button(__("Update Commission"), function () {
      console.log("Apply Commission button clicked ::::");
      let by_percent_commission =
        report.page.fields_dict.by_percent_commission.get_value();
      let by_amount_commission =
        report.page.fields_dict.by_amount_commission.get_value();

      if (!by_percent_commission && !by_amount_commission) {
        frappe.msgprint(
          __("Please enter either a commission percentage or an amount.")
        );
        return;
      }

      let docnames = report.data.map((row) => row.docname).filter(Boolean);

      frappe.call({
        method: "sranco.api.update_item_commissions",
        args: {
          docnames: JSON.stringify(docnames),
          by_percent: by_percent_commission,
          by_amount: by_amount_commission,
        },
        callback: function (response) {
          if (response.message === "success") {
            console.log("Commissions are updated successfully ::: ");
            frappe.msgprint(__("Commissions updated successfully."));
            report.refresh(); // Reload report to show updated commissions
          } else {
            frappe.msgprint(__("Error updating commissions."));
          }
        },
      });
    });
  },
  filters: [
    {
      fieldname: "from_date",
      label: __("From Date"),
      fieldtype: "Date",
      default: frappe.datetime.add_months(frappe.datetime.get_today(), -1),
      reqd: 1,
    },
    {
      fieldname: "to_date",
      label: __("To Date"),
      fieldtype: "Date",
      default: frappe.datetime.get_today(),
      reqd: 1,
    },
    {
      fieldname: "item_code",
      label: __("Item"),
      fieldtype: "Link",
      options: "Item",
    },
    {
      fieldname: "item_name",
      label: __("Item"),
      fieldtype: "Link",
      options: "Item",
    },
    {
      fieldname: "tn_number",
      label: __("TN Number"),
      fieldtype: "Data",
    },
    {
      fieldname: "customer",
      label: __("Customer"),
      fieldtype: "Link",
      options: "Customer",
    },
  ],
};

File: sranco/sranco/report/snc_commision_history/snc_commision_history.py
# Copyright (c) 2023, Dinesh Panchal and contributors
# For license information, please see license.txt

import frappe
from frappe.utils import logger

logger.set_log_level("DEBUG")
logger = frappe.logger("Sranco_logs", allow_site=True, file_count=1)

def execute(filters=None):
    columns, data = get_columns_and_data_commission(filters)
    return columns, data

def get_columns_and_data_commission(filters):
    columns = [
        {
            "fieldname": "item_code",
            "label": "Item Code",
            "fieldtype": "Link",
            "options": "Item",
            "width": 200
        },
        {
            "fieldname": "item_name",
            "label": "Item Name",
            "fieldtype": "Data",
            "width": 200
        },
        {
            "fieldname": "commission_percent",
            "label": "Commission %",
            "fieldtype": "Percent",
            "width": 120
        },
        {
            "fieldname": "commission_amount",
            "label": "Commission Amount",
            "fieldtype": "Currency",
            "width": 130
        },
        {
            "fieldname": "changed_by",
            "label": "Changed By",
            "fieldtype": "Data",
            "width": 150
        },
        {
            "fieldname": "date_time",
            "label": "Date and Time",
            "fieldtype": "Datetime",
            "width": 200
        },
        {
            "fieldname": "customer",
            "label": "Customer",
            "fieldtype": "Link",
            "options": "Customer",
            "width": 200
        }
    ]

# ... (Continuing from the previous part)

    item_data = get_item_details_commission(filters)
    data = []

    for item in item_data:
        current_commission_percent = item.get("current_commission_percent")
        current_commission_amount = item.get("current_commission_amount")
        doc = item.get("docname")
        itemCode = item.get("item_code")
        itemName = item.get("item_name")
        customer = item.get("customer")
        
        row = {
            "docname": doc,
            "item_code": itemCode,
            "item_name": itemName,
            "commission_percent": current_commission_percent,
            "commission_amount": current_commission_amount,
            "changed_by": f"<b>Current Commission</b>",
            "date_time": None,
            "customer": customer,
        }
        data.append(row)

        # Fetch the commission history for the current item
        commission_history_data = get_commission_history_for_item(item.get("docname"), filters)

        for history in commission_history_data:
            row = {
                "docname": "",
                "item_code": "",
                "item_name": "",
                "commission_percent": history.get("percent"),
                "commission_amount": history.get("amount"),
                "changed_by": history.get("changed_by"),
                "date_time": history.get("date_time"),
                "customer": ""
            }
            data.append(row)

    return columns, data

def get_item_details_commission(filters):
    conditions = []

    if filters.get("tn_number"):
        conditions.append("i.custom_tn_number = %(tn_number)s")
    if filters.get("customer"):
        conditions.append("ip.customer = %(customer)s")
    if filters.get("item_code"):
        conditions.append("ip.item_code = %(item_code)s")

    conditions = " AND ".join(conditions)

    return frappe.db.sql("""
        SELECT DISTINCT
            ip.name as docname,
            ip.item_code,
            ip.item_name,
            ip.customer,
            ip.custom_snc_commission_ as current_commission_percent,
            ip.custom_snc_commission_amount as current_commission_amount
        FROM
            `tabItem Price` ip
        JOIN `tabItem` i ON ip.item_code = i.name
        WHERE {conditions}
    """.format(conditions=conditions or "1=1"), filters, as_dict=1)

def get_commission_history_for_item(docName, filters):
    conditions = ["ip.name = %(docName)s"]

    # Format the dates with time
    if filters.get("from_date"):
        filters["from_datetime"] = filters["from_date"] + " 00:00:00"
        conditions.append("ch.date_time >= %(from_datetime)s")
    if filters.get("to_date"):
        filters["to_datetime"] = filters["to_date"] + " 23:59:59"
        conditions.append("ch.date_time <= %(to_datetime)s")
    if filters.get("item_code"):
        conditions.append("ip.item_code = %(item_code)s")

    conditions = " AND ".join(conditions)
    
    return frappe.db.sql("""
        SELECT
            ch.percent,
            ch.amount,
            ch.changed_by,
            ch.date_time
        FROM
            `tabItem Price` ip
        LEFT JOIN `tabSNC Commission History` ch ON ip.name = ch.parent
        WHERE {conditions}
        ORDER BY ch.date_time DESC
    """.format(conditions=conditions), {**filters, "docName": docName}, as_dict=1)
